README.md
[![Build Status](https://dev.azure.com/lazywinadmin/AdsiPS/_apis/build/status/lazywinadmin.AdsiPS?branchName=master)](https://dev.azure.com/lazywinadmin/AdsiPS/_build/latest?definitionId=17&branchName=master)

# AdsiPS

PowerShell module to interact with Active Directory using ADSI and the System.DirectoryServices namespace (.NET Framework)

The initial motivation for this module was to improve my knowledge on how to interact with Active Directory without the Microsoft Active Directory module or the Quest Active Directory Snapin.
The other elements that I wanted to work on were being able to use alternative Credentials and to specify a different Domain.

Obviously I'm still learning and there is ton of space for improvements... Would love contributors, suggestions, feedback or any other help.

## Table of contents

- [Contributing](#contributing)
- [Installation](#installation)
- [Download from PowerShell Gallery](#Download-from-PowerShell-Gallery)
- [Download from GitHub repository](#Download-from-GitHub-repository)
- [Use Cases](#use-cases)
- [More Information](#more-information)
- [Notes](#notes)

## Contributing

Contributions are welcome via pull requests and issues.
Please see our [contributing guide](https://github.com/lazywinadmin/adsips/blob/master/CONTRIBUTING.md) for more details

## Installation

### Download from PowerShell Gallery (recommended)

Only from PowerShell version 5

``` powershell
Install-Module -name ADSIPS
```

### Download from GitHub repository

1. Download the repository
1. Unblock the zip file
1. Extract the folder to a module path (e.g. $home\Documents\WindowsPowerShell\Modules)
1. Run `build.ps1` (exists in project root).
   - **NOTE:** If you get an error after running `build.ps1` - please use **`build.ps1 -InstallDependencies`**
1. `build.ps1` creates a folder called `~\buildoutput\AdsiPs` in the directory which `AdsiPs` was saved to
1. Inside of `\buildoutput\AdsiPs` there is a file called `AdsiPs.psm1`
1. Run `Import-Module -Path "C:\Path\To\buildoutput\AdsiPs\AdsiPs.psm1"` to import the `AdsiPs` module

## Use Cases

1. Learning Active Directory: We can't see the code behind the Microsoft ActiveDirectory Module and Quest ActiveDirectory Snapin. This module is a great way to explore and learn on how Active Directory is working,
1. Delegation: Active Directory queries need to be performed by a tool (GUI for example) and you don't want it to load AD module. Additionally you don't know who will use the tool and if they have/can/know how to install the module,
1. Performance:  ADSI is way faster,
1. Restricted environment: Sometime ActiveDirectory Module is not available/ or can't install it on a machine.

## More Information

* MSDN is a great resource if you want to find more information on the NET classes to use. See [System.DirectoryServices](https://msdn.microsoft.com/en-us/library/system.directoryservices(v=vs.110).aspx)

## Authors

* Thanks to our Contributors!!
  * @MickyBalladelli
  * @christophekumor
  * @omiossec
  * @oze4
  * @andrewtchilds
  * @NicolasBn
  * @gerane

## Resources

Interesting projects using different approaches to reach out to Active Directory in PowerShell/c#
* [PSAD by @zloeber](https://github.com/zloeber/PSAD)
* [ADAudit by @darkoperator](https://github.com/darkoperator/ADAudit/tree/dev)
* [ADSI on powershell.com by TobiasPSP](http://powershell.com/cs/blogs/ebookv2/archive/2012/03/25/chapter-19-user-management.aspx) by Tobias Weltner
* [ADRecon from @sense-of-security](https://github.com/sense-of-security/ADRecon)
* [PowerView from @PowerShellMafia team](https://github.com/PowerShellMafia/PowerSploit/blob/master/Recon/PowerView.ps1)
* [Invoke-Kerberoast from @EmpireProject team](https://github.com/EmpireProject/Empire/blob/master/data/module_source/credentials/Invoke-Kerberoast.ps1)
* [Test-ActiveDirectory by markwragg](https://github.com/markwragg/Test-ActiveDirectory/blob/master/ADAudit/ActiveDirectory.tests.ps1)
* [AdEnumerator(LDAP) by @chango77747](https://github.com/chango77747/AdEnumerator/blob/master/ADEnumerator.psm1)
* [Grant-ADPermission by @edemilliere](https://github.com/edemilliere/ADSI/blob/master/Grant-ADPermission.ps1)



CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [1.0.0.6] - 2019-06-23
### Added
- Add Get-ADSIUserPrimaryGroup (by @oze4)
  - This was added as part of the resolution for issue #70
  - By putting this function in it's own ps1 file, it keeps things clean and organized
  - Provides an additional 'feature' (is not limited to helping fix issue #70 - you can now get a users primary group very easily)

### Changed
- Update Readme.md (by @oze4)
  - Numbering was not correct in regards to the last changes made
  - Cleaned it up
- Update Build.ps1 (by @oze4)
  - Default tasks: Build
  - Add Warning message for Dependencies
- Update Get-ADSIPrincipalGroupMembership (by @oze4)
  - Fix indentation
  - Fix #70 "Get-ADSIPrincipalGroupMembership Problem since Last update"
    - This was due to Primary User Group not being added to overall Group Membership
- Update Get-ADSIUser (by @oze4)
  - Fix #72 "Review ParameterSetName to allow lazy typing"

## [1.0.0.5] - 2019-06-21
### Changed
-Update Pester test to receive srcpath and compare source ps1 files and build functions
-Update Manifest try,catch blocks

## [1.0.0.4] - 2019-06-21
### Changed
-Build as been moved to Azure DevOps/Azure Pipelines and now auto-deploy to PowerShell Gallery

## [1.0.0.3] - 2017-11-30

## [1.0.0.2] - 2016-08-20

## [1.0.0.1] - 2016-08-17

build.ps1
<#
.SYNOPSIS
Used to start the build of a PowerShell Module
.DESCRIPTION
This script will install dependencies using PSDepend module and start the build tasks using InvokeBuild module
.NOTES
Change History
-1.0 | 2019/06/17 | Francois-Xavier Cat (lazywinadmin)
    - Initial version
-1.1 | 2019/06/22 | Matt Oestreich (oze4)
    - Added specific error handling/error message in regards to missing dependencies and how to resolve them
    - Specifically, to use the `-InstallDependencies` switch with `build.ps1`
    - Removed redundant `[string[]]$tasks` parameter.. (param appears to be in use, but there was a duplicate, which was commented out, so I just removed it)
    - Added 'tasks' param check so if user wants to build this locally, they don't have to know to supply the 'tasks' param with a value of @('build')
        - TODO: rewrite this using switch params
-1.2 | 2019/06/22 | Francois-Xavier Cat (lazywinadmin)
    - Make build default value of -tasks
    - Move -tasks as first param
    - Remove Error handling around Invoke-Build, this can cause misleading error if it's unrelated to module dependencies
        For example I got an error "Missing dependencies" because it failed on a Pester Test
    - Added Warning message in the global CATCH related to dependencies
#>
[CmdletBinding()]
Param(
    [string[]]$tasks= @('build'),
    [string]$GalleryRepository,
    [pscredential]$GalleryCredential,
    [string]$GalleryProxy,
    [switch]$InstallDependencies
    )
try{
    ################
    # EDIT THIS PART
    $moduleName = "AdsiPS" # get from source control or module ?
    $author = 'Francois-Xavier Cat' # fetch from source or module
    $description = 'PowerShell module to interact with Active Directory using ADSI and the System.DirectoryServices namespace (.NET Framework)' # fetch from module ?
    $companyName = 'lazywinadmin.com' # fetch from module ?
    $projectUri = "https://github.com/lazywinadmin/$moduleName" # get from module of from source control, env var
    $licenseUri = "https://github.com/lazywinadmin/$moduleName/blob/master/LICENSE.md"
    $tags = @('ADSI', 'ActiveDirectory','LDAP','AD')
    ################

    #$rootpath = Split-Path -path $PSScriptRoot -parent
    $rootpath = $PSScriptRoot
    $buildOutputPath = "$rootpath\buildoutput"
    $buildPath = "$rootpath\build"
    $srcPath = "$rootpath\src"
    $testPath = "$rootpath\tests"
    $modulePath = "$buildoutputPath\$moduleName"
    $dependenciesPath = "$rootpath\dependencies" # folder to store modules
    $testResult = "Test-Results.xml"

    $env:moduleName = $moduleName
    $env:modulePath = $modulePath

    $requirementsFilePath = "$buildPath\requirements.psd1" # contains dependencies
    $buildTasksFilePath = "$buildPath\tasks.build.ps1" # contains tasks to execute

    if($InstallDependencies)
    {
        # Setup PowerShell Gallery as PSrepository  & Install PSDepend module
        if (-not(Get-PackageProvider -Name NuGet -ForceBootstrap)) {
            $providerBootstrapParams = @{
                Name = 'nuget'
                force = $true
                ForceBootstrap = $true
            }

            if($PSBoundParameters['verbose']) {$providerBootstrapParams.add('verbose',$verbose)}
            if($GalleryProxy) { $providerBootstrapParams.Add('Proxy',$GalleryProxy) }
            $null = Install-PackageProvider @providerBootstrapParams
            Set-PSRepository -Name PSGallery -InstallationPolicy Trusted
        }

        if (-not(Get-Module -Listavailable -Name PSDepend)) {
            Write-verbose "BootStrapping PSDepend"
            "Parameter $buildOutputPath"| Write-verbose
            $InstallPSDependParams = @{
                Name = 'PSDepend'
                AllowClobber = $true
                Confirm = $false
                Force = $true
                Scope = 'CurrentUser'
            }
            if($PSBoundParameters['verbose']) { $InstallPSDependParams.add('verbose',$verbose)}
            if ($GalleryRepository) { $InstallPSDependParams.Add('Repository',$GalleryRepository) }
            if ($GalleryProxy)      { $InstallPSDependParams.Add('Proxy',$GalleryProxy) }
            if ($GalleryCredential) { $InstallPSDependParams.Add('ProxyCredential',$GalleryCredential) }
            Install-Module @InstallPSDependParams
        }

        # Install module dependencies with PSDepend
        $PSDependParams = @{
            Force = $true
            Path = $requirementsFilePath
        }
        if($PSBoundParameters['verbose']) { $PSDependParams.add('verbose',$verbose)}
        Invoke-PSDepend @PSDependParams -Target $dependenciesPath
        Write-Verbose -Message "Project Bootstrapped"
    }

    # Start build using InvokeBuild module
    Write-Verbose -Message "Start Build (using InvokeBuild module)"
    Invoke-Build -Result 'Result' -File $buildTasksFilePath -Task $tasks


    # Return error to CI
    if ($Result.Error)
    {
        $Error[-1].ScriptStackTrace | Out-String
        exit 1
    }
    exit 0
}catch{
    Write-Warning -Message "It is possible you are missing dependencies. Please rerun using 'build.ps1 -InstallDependencies' if you do not have the 'PSDepend' module installed!"
    throw $_
}
